Add support for "write-once" properties

kocsismate opened the pull request:
Currently, the implementation makes use of the `final` keyword, although the actual name is to be decided.

RFC: https://wiki.php.net/rfc/write_once_properties

HallofFamer commented:
I think this is a good idea, and the keyword **final** is quite appropriate here. Java has final class/method/property, while C# doesnt have the final keyword and instead use other keywords like 'sealed' and 'readonly'. From the aspect of consistency, the keyword final is very suitable for immutable properties too.

kocsismate commented:
@HallofFamer Unfortunately, I won't propose final as a keyword in the RFC :(

In PHP, `final` currently prevents a class or a method to be overridden. In the sense of this, using `final` for "immutable" properties would be a mistake.

nikic commented:
This doesn't implement the "references are not allowed" part yet, right?

kocsismate commented:
I would say I only had a try at it, but haven't looked at it again since then... https://github.com/php/php-src/blob/6513e68b7f2dd9aee31dce2667da376931a0ae06/Zend/zend_object_handlers.c#L1054

mart-k commented:
Maybe I missed it in RFC, but what happens when you call next($this->c); ? Will it move the internal pointer (of an array) or throw an exception?

kocsismate commented:
@mart-k I've just done a quick test, and actually, `next()` throws an exception. :( I think it should work since it's needed for reading, so I'll add a test case for this and then try to fix the issue. Also I'll modify the RFC to mention the expected behaviour. Thank you for pointing this out!

nikic commented:
@kocsismate Err, why would next() work? It modifies the array, so it should not be permitted.

Note that foreach(), ArrayIterator etc all work independently of the internal array pointer and can work fine with a readonly property.

kocsismate commented:
> Note that foreach(), ArrayIterator etc all work independently of the internal array pointer and can work fine with a readonly property.

Oh, I thought that the internal pointer is used during iteration, so my reasoning was that then `next()` should be allowed. Sorry for this. But the testcase + another example come in handy indeed.

mart-k commented:
One of the reasons why I pointed next() out is that there are whole family of similar functions that read/write array's "internal pointer". For example what happens if you call current()? It is read only and doesn't change the pointer, but if next(), each() etc. give error, will allowing current() even make sense?

kocsismate commented:
@mart-k Calling `current()` currently works. My opinion is that it doesn't make sense to disallow those functions that only read the pointer... But Nikita should be more informed.

nikic commented:
@kocsismate Right, I don't see why these need any special consideration. They are normal by-value or by-reference functions, depending on whether they perform modification or not.

someniatko commented:
What about clonable objects? There is a popular `#withX()` pattern which first clones an object, and then assigns the prop of cloned object new value, therefore preserving immutability. It is used e.g. in PSR-7 interfaces (methods like `withHeader()`).

Like this:

```php
class C
{
    writeonce public int $prop;
    
    public function __construct(int $prop)
    {
        $this->prop = $prop;
    }

    public function withProp(int $prop)
    {
         $that = clone $this;
         $that->prop = $prop; // will fail?
    }
}
```

kocsismate commented:
@someniatko Cloning is currently under debate. My point of view is that it's not a prerequisite for writeonce/readonly properties since you can create a completely new instance instead of using `clone`. It's not that comfortable though, but I'd like to come up with a way to support mutability of immutable properties. :)

AnrDaemon commented:
IMO, this suggestion is extremely unsettling. Probably on the level of singletons - it produces a sudden behavior change that can't be explained by surrounding code logic.

If you want special actions taken for writing data to the property, a more generic approach can be taken.

Something like

```php
[static] property [typehint] $name [write methodName];
```

where `methodName` is a class method with signature identical to `__set()` except for matching typehints. F.e. `function(string $name, ?typehint $value)`.
If no setter method declared, it will be public read-only property. If a setter is declared, an attempt to write to the property outside the object/class will be redirected to the setter.

It is arguable if a setter requires `$name` in its arguments, but I think this can increase usability of the feature by allowing for setter reuse without writing a lot of proxy methods.

Aaronidas commented:
What would happen with property reflection? I haven't read about reflection in the RFC, I guess it should be taken into account and add a method to identify whether a property is immutable or not.

kocsismate commented:
@Aaronidas yeah, I'll definitely add a new `ReflectionProperty::isReadonly()` method. It's also worth a mention in the RFC. Thanks for the reminder.

someniatko commented:
@kocsicmate i think you should be consistent and name it `isWriteonce()`, please.

kocsismate commented:
@someniatko sorry, I was a bit unclear: the actual method name depends on the outcome of the secondary vote which is about the keyword choice.

nikic commented on Zend/tests/final_properties/access/property_incrementing_decrementing_by_ref.phpt:
```diff
@@ -0,0 +1,80 @@
+
+?>
+--EXPECT--
+Cannot increment a reference held by final property Foo::$property1
```
That means it was possible to acquire a reference to it in the first place...

kocsismate replied:
Uh, yes.. However, It's only because I couldn't prevent references in case of `foreach ($foo as &$property) {}`, while other ways of acquiring references doesn't seem to work anymore.

kocsismate replied:
Ah, I see that it was a question in your main comment. I simply didn't find find out how to deal with foreach-by-ref. Can you tell me where I should be searching?

nikic replied:
You'll want to insert a check somewhere around here: https://github.com/php/php-src/blob/191451da90b6fe81641d55bf782b551c65e4fd70/Zend/zend_vm_def.h#L6516

nikic commented on Zend/tests/final_properties/clone/clone_error.phpt:
```diff
@@ -0,0 +1,38 @@
+--TEST--
+Test that final properties can't be modified after cloning
```
I assume cloning in the uninitialized state and assigning then would work though, right?

kocsismate replied:
Right! I'll add a test for this case in `clone_success.php`, I was indeed missed out.

nikic commented on Zend/tests/final_properties/inheritance/inheritance_success1.phpt:
```diff
@@ -0,0 +1,19 @@
+class Bar extends Foo
+{
+    protected static int $property1;
+    protected int $property2;
```
I don't think the inheritance behavior is mentioned in the RFC yet. Dropping final does seem sound though.

I was also wondering if we could make the type covariant for readonly properties. I.e. `final iterable $iter` to `final array $iter`. Technically this is not quite sound because it restricts the type of the initial assignment...

kocsismate replied:
Unfortunately, I neglected the inheritance aspect because I concentrated too much on making runtime restrictions work... So I'll add a related paragraph to the RFC.

Would property covariance be special to readonly properties? For BC reasons maybe? (properties are now invariant, right?)

nikic commented on Zend/tests/final_properties/inheritance/inheritance_success2.phpt:
```diff
@@ -0,0 +1,19 @@
+class Bar extends Foo
+{
+    final protected static int $property1;
+    final protected int $property2;
```
This looks unsound. Foo may be assuming that it can write to the properties.

kocsismate replied:
Oh, you are right of course.

nikic commented on Zend/tests/final_properties/unset/declared_unset_error.phpt:
```diff
@@ -0,0 +1,39 @@
+    echo $exception->getMessage() . "\n";
+}
+
+// Uninitialized typed properties can't be unset
```
can't -> can?

kocsismate replied:
It is just an outdated text before I discovered how unsetting actually works. ^^ I'm deleting it, no need for it now.

nikic commented on Zend/tests/final_properties/unset/magic_unset_success.phpt:
```diff
@@ -0,0 +1,41 @@
+--TEST--
+Test that the __unset() magic method is successful
+--XFAIL--
+Unsetting is currently not possible because of "Cannot modify final property Foo::$properties by reference after initialization"
```
Not seeing a reference here, but the fact that it's forbidden is right, no?

kocsismate replied:
Ideally, reference shouldn't be involved, so I'll investigate what happens here. Besides, my reasoning is that `$properties` is not initialized, so unsetting it (via `__unset`) should be possible. Or am I missing something?

nikic replied:
Yeah, you're right, I didn't notice that it's not initialized.

kocsismate commented:
@nikic My newest commit addresses some of your concerns, like adding the missing tests and inheritance checks. Thanks to your help I think I managed to make the "foreach-by-ref"-related tests work.

However, lazy initialization didn't yet work because `zend_verify_prop_assignable_by_ref()` is called when calling getters and I implemented the "Cannot modify final properties by reference" check there. I am not sure if I did the right thing, but I disabled this validation in this very case, and now it's all good... I guess the `magic_unset_success.phpt` test doesn't work because of a similar reason (and that's why it says reference while in fact it isn't involved).

nikic commented on Zend/tests/final_properties/access/property_by_ref_access3.phpt:
```diff
@@ -0,0 +1,32 @@
+?>
+--EXPECTF--
+Notice: Accessing static property Foo::$property1 as non static in %s:%d
+Cannot acquire reference on final static property Foo::$property1
```
I think this test is already working correctly. The error message is not super clear about it, but what happens here is that `$this->property1 = 1` actually creates a dynamic instance property on the object, which is unrelated to the static property declaration and as such does not enforce final.

kocsismate replied:
Oh, nice! I didn't know what happens in this case.

kocsismate replied:
Nope. I didn't pay attention to `$this->property1 = 1`... And I wanted to use `self::$property1 = 1` Then it's all clear, I just didn't double checked it.

nikic commented on Zend/zend_execute.c:
```diff
@@ -3134,12 +3210,21 @@ ZEND_API zval* zend_assign_to_typed_ref(zval *variable_ptr, zval *value, zend_uc
 
 		val = Z_REFVAL_P(val);
+
+		if (check_final &&prop_info && prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(val) != IS_PROP_UNINIT) {
```
```suggestion
		if (check_final && prop_info && prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(val) != IS_PROP_UNINIT) {
```

nikic replied:
I don't think the `IS_PROP_UNINIT` check here makes sense. `val` is the value you're assigning (not the property value), so it does not support this flag (it should be uninitialized memory in fact).

kocsismate replied:
I should have noticed it :) And the check doesn't make sense because even non-initialized properties shouldn't be assigned by ref.

Then I'll do the this validation in `zend_assign_to_typed_property_reference()` and that way at least we don't need the `check_final` flag anymore.

nikic commented on Zend/zend_compile.c:
```diff
@@ -6386,19 +6386,18 @@ void zend_compile_prop_decl(zend_ast *ast, zend_ast *type_ast, uint32_t flags) /
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Final property %s::$%s must have a type",
+        		ZSTR_VAL(ce->name), ZSTR_VAL(name)
+        	);
```
Some space indentation snuck in here.

nikic commented on Zend/zend_execute.c:
```diff
@@ -1655,7 +1663,7 @@ static void zend_incdec_typed_ref(zend_reference *ref, zval *copy OPLINE_DC EXEC
 
 	if (UNEXPECTED(Z_TYPE_P(var_ptr) == IS_DOUBLE) && Z_TYPE_P(copy) == IS_LONG) {
-		zend_property_info *error_prop = zend_get_prop_not_accepting_double(ref);
+		zend_property_info *error_prop = error_prop = zend_get_prop_not_accepting_double(ref);
```
I don't think you intended to do that :)

kocsismate replied:
😵

nikic commented on Zend/zend_execute.c:
```diff
@@ -1685,7 +1693,12 @@ static void zend_incdec_typed_prop(zend_property_info *prop_info, zval *var_ptr,
 	}
 
-	if (UNEXPECTED(Z_TYPE_P(var_ptr) == IS_DOUBLE) && Z_TYPE_P(copy) == IS_LONG) {
+	if (UNEXPECTED(prop_info->flags & ZEND_ACC_FINAL)) {
```
You can move this check earlier and save some of the copy juggling. The other checks occur after because they compare the old and new vals.

nikic commented on Zend/zend_execute.c:
```diff
@@ -2629,6 +2679,15 @@ static zend_never_inline zend_bool zend_handle_fetch_obj_flags(
 				}
+
+				if (prop_info->flags & ZEND_ACC_FINAL) {
+					zend_throw_error(NULL, "Cannot acquire reference on final property %s::$%s",
```
I'd use "reference to" rather than "reference on".

nikic commented on Zend/zend_execute.c:
```diff
@@ -3160,6 +3245,14 @@ ZEND_API zend_bool ZEND_FASTCALL zend_verify_prop_assignable_by_ref(zend_propert
 	} else {
 		ZVAL_DEREF(val);
+
+		if (check_final &&prop_info && prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(val) != IS_PROP_UNINIT) {
```
Looks like the code is the same in both branches, can be hoisted.

nikic commented on Zend/zend_object_handlers.c:
```diff
@@ -805,7 +817,13 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
+
 		if (Z_TYPE_P(variable_ptr) != IS_UNDEF) {
+			if (UNEXPECTED(Z_PROP_FLAG_P(variable_ptr) != IS_PROP_UNINIT && prop_info && prop_info->flags & ZEND_ACC_FINAL)) {
+				zend_final_property_assignment_error(zobj->ce, name);
```
Needs a `variable_ptr = &EG(error_zval);` I think.

nikic commented on Zend/zend_object_handlers.c:
```diff
@@ -1502,6 +1532,14 @@ ZEND_API zval *zend_std_get_static_property_with_info(zend_class_entry *ce, zend
 	ret = CE_STATIC_MEMBERS(ce) + property_info->offset;
 	ZVAL_DEINDIRECT(ret);
 
+	if (UNEXPECTED((type == BP_VAR_RW || type == BP_VAR_W) && property_info->flags & ZEND_ACC_FINAL && Z_TYPE_P(ret) != IS_UNDEF)) {
```
Hm... I think there's a pretty big problem here :(

We currently compile the operands of property accesses with the property access mode, i.e. `Foo::$bar->baz = 1` is going to perform a W fetch on the static property, rather than the R fetch you would expect.

This means that code like this is going to error, while it should not:

```
<?php
class Foo {
    final public static stdClass $prop;
}

Foo::$prop = new stdClass;
Foo::$prop->bar = 1;
```

Probably more test coverage for modification of objects stored in readonly properties is needed.

nikic replied:
I tried to fix this in #3865, but ran into some compatibility issues with SimpleXML.

kocsismate replied:
Sounds really bad. :/ But I'll add more tests of course. Do you have any idea if this issue is a deal breaker or could be somehow circumvented? Is it worth to have SimpleXML in core if it has issues like this?

nikic replied:
For the record, I opened #5250 and started a thread at https://externals.io/message/108950.

> Btw is it worth to have SimpleXML in core if it has issues like this?

It's not, but people will complain if we drop it :)

nikic commented on Zend/zend_execute.c:
```diff
@@ -2642,6 +2708,23 @@ static zend_never_inline zend_bool zend_handle_fetch_obj_flags(
+				if (result) ZVAL_ERROR(result);
+				return 0;
+			}
+			break;
```
Does this code actually do anything? I don't see any corresponding compiler code that sets these flags, so I would expect it to never run.

kocsismate replied:
No, not much. I put it as an earlier experimentation (without checking the compiler), but forgot to remove it.

kocsismate replied:
Hmm, or is it possible that I'll need this code? Currently, I'm trying to implement the restriction for the `ZEND_FETCH_OBJ_UNSET` opcode.  I see that `zend_fetch_property_address()` is called from the opcode handler with `flags=0`. However, should I have this piece of code and set the flag then maybe `unset($foo->a[0])` could be checked? Or is there a better way?

nikic replied:
Yes, I think you're going to need at least the UNSET portion of this, in order to distinguish between `unset($foo->a[0])` and `unset($foo->a->b)` (otherwise you could just check for BP_VAR_UNSET). You'll have to add the corresponding flag in the compiler (like it is done for the two flags that are checked above). One value was freed up in PHP 8, so it should fit in.

nikic commented:
I'm missing some tests for `$obj->prop += 1` or similar. I think it might not be handled right now.

kocsismate commented on Zend/tests/final_properties/access/property_assignment1.phpt:
```diff
@@ -0,0 +1,104 @@
+--TEST--
+Test that final properties can't be mutated by assignment operators
```
These tests might not be very useful because all of them boil down to simple assignments from the perspective of "write-once" properties

kocsismate commented on Zend/tests/final_properties/access/property_assignment_combined1.phpt:
```diff
@@ -0,0 +1,106 @@
+}
+
+try {
+    $foo->property2 ??= "foo";
```
I haven't managed to make this case work.

nikic replied:
This only does an assignment if the property is null. So it's correct that this doesn't error as-is, you'd have to use a different default value.

kocsismate replied:
What a little bastard :'D But thanks for the info, it was too difficult for me to discover how it really works.

nikic commented:
One more thing to test would be by-ref passing. For example `sort($obj->prop)`.

kocsismate commented on Zend/zend_vm_def.h:
```diff
@@ -1804,7 +1804,7 @@ ZEND_VM_HELPER(zend_fetch_static_prop_helper, ANY, ANY, int type)
 	SAVE_OPLINE();
 
-	if (UNEXPECTED(zend_fetch_static_property_address(&prop, NULL, opline->extended_value & ~ZEND_FETCH_OBJ_FLAGS, type, opline->extended_value & ZEND_FETCH_OBJ_FLAGS OPLINE_CC EXECUTE_DATA_CC) != SUCCESS)) {
+	if (UNEXPECTED(zend_fetch_static_property_address(&prop, NULL, opline->extended_value & ~(type == BP_VAR_UNSET ? ZEND_FETCH_DIM_UNSET_FLAG : ZEND_FETCH_OBJ_FLAGS), type, opline->extended_value & (type == BP_VAR_UNSET ? ZEND_FETCH_DIM_UNSET_FLAG : ZEND_FETCH_OBJ_FLAGS) OPLINE_CC EXECUTE_DATA_CC) != SUCCESS)) {
```
It doesn't feel the right solution for me, but I didn't have much clue what to do here

TysonAndre commented on Zend/zend_object_handlers.c:
```diff
@@ -805,7 +817,14 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
 		variable_ptr = OBJ_PROP(zobj, property_offset);
+
 		if (Z_TYPE_P(variable_ptr) != IS_UNDEF) {
+			if (UNEXPECTED(Z_PROP_FLAG_P(variable_ptr) != IS_PROP_UNINIT && prop_info && prop_info->flags & ZEND_ACC_FINAL)) {
```
I think this could just be `ZEND_ASSERT(prop_info)` before this (only for `IS_VALID_PROPERTY_OFFSET`). The prop_info should be set if a valid property reference was null, if I understand how this works correctly.

I'd guess it'd be more efficient as `UNEXPECTED(prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(variable_ptr) != IS_PROP_UNINIT)`, because ACC_FINAL would probably be much less common than being initialized.

```
#define IS_VALID_PROPERTY_OFFSET(offset)           ((intptr_t)(offset) > 0)
#define IS_WRONG_PROPERTY_OFFSET(offset)           ((intptr_t)(offset) == 0)
#define IS_DYNAMIC_PROPERTY_OFFSET(offset)         ((intptr_t)(offset) < 0)
```

kocsismate replied:
I'll change the order of the condition, thank you for the pointing this out!

However, it seems to me that it's not possible to add a `ZEND_ASSERT(prop_info)` because as far as I saw, it's only set for typed properties. And the `if (UNEXPECTED(prop_info))` condition some lines below also confirm this I think.

TysonAndre commented on Zend/tests/final_properties/access/property_assignment_combined1.phpt:
```diff
@@ -0,0 +1,106 @@
+    echo $exception->getMessage() . "\n";
+}
+
+$foo->property2 ??= null;
```
Could document in the source of the test case that this is the first assignment to property2, and the test chose `null` because `??=` would do nothing if the property value was not null and not initialized - I assume others may be as confused as you.

Same for the tests of static.

kocsismate replied:
Will do! Actually, I was already considering it before, but then changed my mind.

TysonAndre commented:
Other edge cases that might need extra code in the unserializer, or warrant tests (I'd definitely think they'd be fixable if they have bugs):

- Malformed data which sets the final property twice should throw (probably already handled?) (`O:7:"MyClass":2:{s:9:"finalProp";i:2;s:9:"finalProp";i:3;}`)

  An annoying edge case would be unserializing properties that already have defaults - maybe those objects shouldn't throw at all (e.g. in `__wakeup` or `__unserialize`) until unserialization is over.
- The unserializer creating final properties as references (e.g. due to being serialized with an older class definition which didn't use `final`)
  `'O:7:"MyClass":2:{s:9:"finalProp";i:2;s:7:"refProp";R:2;}'` for `$x->refProp = &$x->finalProp;`

I can imagine serialize() being used with final properties (e.g. to load from the database once and cache in memcache).

I'd naively guess the below snippet might throw an exception the way this is currently implemented, and if it doesn't, it'd still be useful to have the test to ensure behavior doesn't accidentally change in the future.

```php
<?php

class X {
    final public $hasDefault = ['v7'];
}
$x = new X();
$s = serialize($x);
// Desired: This should not throw.
var_dump(unserialize($x));
```

(I maintain https://github.com/igbinary/igbinary , a compact binary serializer for php)

kocsismate commented:
@TysonAndre I've started to add tests for serialization. I can say that
- Malformed data which sets the final property doesn't throw an exception (yet)
- You are right, final properties that have a default value can be unserialized

I'll continue tomorrow :)

kocsismate commented:
@TysonAndre Today's findings:

> The unserializer creating final properties as references (e.g. due to being serialized with an older class definition which didn't use final)

A serialized data like that is currently possible to unserialize. However, these properties can be currently modified via reference. :/.

> Also, are there tests of references to fields of properties (e.g. what is the intended behavior for preg_match('/abc/', 'abcdef', $x->prop['field'])) - I skimmed the tests and didn't see any

Currently, an exception is thrown in this case with a message of "Cannot modify final property Foo::$property1 after initialization". I added a test for this case.

TysonAndre commented:
> A serialized data like that is currently possible to unserialize. However, these properties can be currently modified via reference. :/.

ext/standard/var_unserializer.re has a check for whether it's possible to assign a value to a typed property while unserializing - I assume a similar error should be thrown if it was write-once. Not completely familiar with this part of the code.

```c
		if (UNEXPECTED(info)) {
			if (!zend_verify_prop_assignable_by_ref(info, data, /* strict */ 1)) {
				zval_ptr_dtor(data);
				ZVAL_UNDEF(data);
				zval_dtor(&key);
				goto failure;
			}
			if (Z_ISREF_P(data)) {
				ZEND_REF_ADD_TYPE_SOURCE(Z_REF_P(data), info);
			}
		}
```

TysonAndre commented on Zend/zend_compile.c:
```diff
@@ -6396,19 +6398,18 @@ void zend_compile_prop_decl(zend_ast *ast, zend_ast *type_ast, uint32_t flags) /
 					ZSTR_VAL(ce->name), ZSTR_VAL(name), ZSTR_VAL(str));
 			}
+		} else if (flags & ZEND_ACC_FINAL) {
+			zend_error_noreturn(E_COMPILE_ERROR,
```
I missed this part of the RFC. https://wiki.php.net/rfc/write_once_properties#compile-type_restrictions

Sadly, there's no way to write `final public resource $fileHandle;`, because `resource` isn't a type declaration in PHP (to give the flexibility to switch resources to objects).

I'd wonder if it'd be possible to remove the syntactic restriction internally (for write-once properties) and represent `final` as a property type info with "every type" if there is no explicit type, i.e. `null|bool|int|float|string|object|array|<resource>` (MAY_BE_RESOURCE) - I don't know if that's actually easy to do, though.

- The alternative I have has its own drawbacks of implicitly and unexpectedly using the uninitialized state of typed properties, so enforcing manually adding a type is probably the best approach for now

kocsismate replied:
I think something similar has already been discussed on the mailing list before. That is, implicitly treating all non-typed properties to be `mixed`. I think it would be much simpler and more straightforward to just wait for the `mixed` type instead. (I really cheer for it to reach PHP 8).

TysonAndre commented on Zend/zend_compile.h:
```diff
@@ -209,7 +209,7 @@ typedef struct _zend_oparray_context {
 /*                                                        |     |     |     */
-/* Final class or method                                  |     |     |     */
+/* Final class, property, or method                       |     |     |     */
 #define ZEND_ACC_FINAL                   (1 <<  5) /*  X  |  X  |     |     */
```
Update the visual representation in the table as well?
```c
/* Class, property and method flags                  class|meth.|prop.|const*/	
```

```suggestion
#define ZEND_ACC_FINAL                   (1 <<  5) /*  X  |  X  |  X  |     */
```

kocsismate replied:
I would rather not update it for now because a new flag will be added for sure if the RFC is accepted (unless it's ok if I use the readonly keyword with `ZEND_ACC_FINAL` - but I think it's not).

TysonAndre commented on Zend/zend_compile.h:
```diff
@@ -907,9 +907,10 @@ ZEND_API zend_string *zend_type_to_string(zend_type type);
-#define ZEND_FETCH_OBJ_FLAGS	3
+#define ZEND_FETCH_REF				1
+#define ZEND_FETCH_DIM_WRITE		2
+#define ZEND_FETCH_DIM_UNSET_FLAG	3
```
If this is deliberately the same value as ZEND_FETCH_OBJ_FLAGS, document the reason why?

kocsismate replied:
Ah, this is just because the flag that was removed from the same place a while ago had the same value. I'll have another shot with this flag since I couldn't make it work for me. :)

mvorisek commented:
The RFC seems to be not accepted, do you know why and do you plan a revision/revote?

kocsismate commented:
@mvorisek I believe the main reasons were these: some people would prefer property accessors, and some people disliked the fact that cloning doesn't play well with readonly properties.

I'll certainly try to revisit this for PHP 8.1. :)

psfpro commented:
Old school style
```php
<?php
class Point {
    protected int $x;
    protected int $y;

    public function __construct(int $x, int $y = 0) {
        $this->x = $x;
        $this->y = $y;
    }
}
```
PHP 8.0.0 style
```php
<?php
class Point {
    public function __construct(protected int $x, protected int $y = 0) {
    }
}
```
This is all good but I still need to write getters. And I'm disappointed that I don't see properties in the class description
```php
<?php
class Point {
    public function __construct(protected int $x, protected int $y = 0) {
    }
    public function getX(): int
    {
        return $this->x;
    }
    public function getY(): int
    {
        return $this->y;
    }
}
```
That would be a great option for example if I want to use a class for DTO
```php
<?php
class Point {
    readonly public int $x;
    readonly public int $y;
}
```
Example
```php
<?php
class Point {
    readonly public int $x;
    readonly public int $y;
}

$point = new Point();
$point->x = 1;
$point->y = 1;

$arr = (array)$point;
echo json_encode($arr);
```

@kocsismate I will be glad to see this RFC in 8.1
